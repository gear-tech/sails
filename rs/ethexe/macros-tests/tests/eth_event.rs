use quote::quote;
use sails_macros_core::derive_eth_event;

#[test]
fn eth_event_basic() {
    let input = quote! {
        pub enum MyEvent {
            MyEvent1,
        }
    };
    let result = derive_eth_event(input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn eth_event_indexed() {
    let input = quote! {
        pub enum Events {
            MyEvent1 {
                #[indexed]
                sender: u128,
                #[indexed]
                amount: u128,
                note: String,
            },
            MyEvent2(#[indexed] u128, u128, String),
            MyEvent3,
        }
    };
    let result = derive_eth_event(input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}

#[test]
fn eth_event_sails_rename() {
    let input = quote! {
        #[sails_path(crate = sails_rename)]
        pub enum Events {
            MyEvent1 {
                #[indexed]
                sender: u128,
                #[indexed]
                amount: u128,
                note: String,
            },
            MyEvent2(#[indexed] u128, u128, String),
            MyEvent3,
        }
    };
    let result = derive_eth_event(input).to_string();
    let result = prettyplease::unparse(&syn::parse_str(&result).unwrap());

    insta::assert_snapshot!(result);
}
