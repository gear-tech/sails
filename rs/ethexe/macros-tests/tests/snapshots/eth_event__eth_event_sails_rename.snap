---
source: macros-tests/tests/eth_event.rs
expression: result
---
#[derive(sails_rename::Encode, sails_rename::TypeInfo)]
#[codec(crate = sails_rename::scale_codec)]
#[scale_info(crate = sails_rename::scale_info)]
pub enum Events {
    MyEvent1 {
        /// #[indexed]
        sender: u128,
        /// #[indexed]
        amount: u128,
        note: String,
    },
    MyEvent2(u128, u128, String),
    MyEvent3,
}
impl sails_rename::EthEvent for Events {
    const SIGNATURES: &'static [sails_rename::gstd::EthEventExpo] = &[
        (
            "MyEvent1",
            <<(
                u128,
                u128,
                String,
            ) as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME,
            sails_rename::keccak_const::Keccak256::new()
                .update("MyEvent1".as_bytes())
                .update(
                    <<(
                        u128,
                        u128,
                        String,
                    ) as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME
                        .as_bytes(),
                )
                .finalize(),
        ),
        (
            "MyEvent2",
            <<(
                u128,
                u128,
                String,
            ) as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME,
            sails_rename::keccak_const::Keccak256::new()
                .update("MyEvent2".as_bytes())
                .update(
                    <<(
                        u128,
                        u128,
                        String,
                    ) as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME
                        .as_bytes(),
                )
                .finalize(),
        ),
        (
            "MyEvent3",
            <<() as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME,
            sails_rename::keccak_const::Keccak256::new()
                .update("MyEvent3".as_bytes())
                .update(
                    <<() as sails_rename::alloy_sol_types::SolValue>::SolType as sails_rename::alloy_sol_types::SolType>::SOL_NAME
                        .as_bytes(),
                )
                .finalize(),
        ),
    ];
    #[allow(unused_variables)]
    fn topics(&self) -> sails_rename::Vec<sails_rename::alloy_primitives::B256> {
        match self {
            Events::MyEvent1 { sender, amount, note } => {
                let mut topics = sails_rename::Vec::with_capacity(3usize);
                let (_, _, hash) = Self::SIGNATURES[0usize];
                topics.push(sails_rename::alloy_primitives::B256::new(hash));
                topics.push(Self::topic_hash(sender));
                topics.push(Self::topic_hash(amount));
                topics
            }
            Events::MyEvent2(f0, f1, f2) => {
                let mut topics = sails_rename::Vec::with_capacity(1usize);
                let (_, _, hash) = Self::SIGNATURES[1usize];
                topics.push(sails_rename::alloy_primitives::B256::new(hash));
                topics
            }
            Events::MyEvent3 => {
                let mut topics = sails_rename::Vec::with_capacity(1usize);
                let (_, _, hash) = Self::SIGNATURES[2usize];
                topics.push(sails_rename::alloy_primitives::B256::new(hash));
                topics
            }
        }
    }
    #[allow(unused_variables)]
    fn data(&self) -> sails_rename::Vec<u8> {
        match self {
            Events::MyEvent1 { sender, amount, note } => Self::encode_sequence(&(note,)),
            Events::MyEvent2(f0, f1, f2) => Self::encode_sequence(&(f0, f1, f2)),
            Events::MyEvent3 => Self::encode_sequence(&()),
        }
    }
}
